app/assets/javascripts/discourse/app/services/media-optimization-worker.js

```diff
@@ -0,0 +1,130 @@
+import Service from "@ember/service";
+import { getOwner } from "@ember/application";
+import { Promise } from "rsvp";
+import { fileToImageData } from "discourse/lib/media-optimization-utils";
+import { getAbsoluteURL, getURLWithCDN } from "discourse-common/lib/get-url";
+
+export default class MediaOptimizationWorkerService extends Service {
+  appEvents = getOwner(this).lookup("service:app-events");
+  worker = null;
+  debugger;
+  workerUrl = getAbsoluteURL("/javascripts/media-optimization-worker.js");
+  currentComposerUploadData = null;
+  currentPromiseResolver = null;
+
+  startWorker() {
+    this.worker = new Worker(this.workerUrl); // TODO come up with a workaround for FF that lacks type: module support
+  }
+
+  stopWorker() {
+    this.worker.terminate();
+    this.worker = null;
+  }
+
+  ensureAvailiableWorker() {
+    if (this.worker === null) {
+      this.startWorker();
+      this.registerMessageHandler();
+      this.appEvents.on("composer:closed", this, "stopWorker");
+    }
+  }
+
+  logIfDebug(message) {
+    if (this.siteSettings.composer_media_optimization_debug_mode) {
+      // eslint-disable-next-line no-console
+      console.log(message);
+    }
+  }
+
+  optimizeImage(data) {
+    let file = data.files[data.index];
+    if (!/(\.|\/)(jpe?g|png|webp)$/i.test(file.type)) {
+      return data;
+    }
+    if (
+      file.size <
+      this.siteSettings
+        .composer_media_optimization_image_kilobytes_optimization_threshold
+    ) {
+      return data;
+    }
+    this.ensureAvailiableWorker();
+    let p = new Promise(async (resolve) => {
```

Comment:
Trivial but you should be able to `return new Promise..` instead of assigning it to a variable and returning that.